drivers: sensor: Added driver for the Würth Elektronik WSEN-HIDS sensor.

yath-eiSmart opened the pull request:
Added driver and sample for the WSEN-HIDS humidity and temperature sensor.

Signed-off-by: Yannick Thesen <Yannick.Thesen@we-online.de>

ycsin commented on drivers/sensor/weplatform/weplatform.c:
```diff
@@ -0,0 +1,134 @@
+			 * Register address most significant bit is used to
+			 * enable multi bytes read
+			 */
+			regAdr |= 1 << 7;
```
[WRITE_BIT](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/util_macro.h#L61)

yath-eiSmart replied:
Done (and moved to hal_we module).

ycsin commented on drivers/sensor/weplatform/weplatform.c:
```diff
@@ -0,0 +1,134 @@
+/*
```
I think it is clearer if you split the i2c & spi like so: https://github.com/zephyrproject-rtos/zephyr/tree/main/drivers/sensor/iis2dh

yath-eiSmart replied:
Done

ycsin commented on drivers/sensor/wsen_hids/Kconfig:
```diff
@@ -0,0 +1,58 @@
+
+config HIDS_ODR
+	string "Output data rate"
+	default "1"
```
would enum be a better choice?

https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/actinius_icarus/dts/bindings/sim-select.yaml#L24

https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/actinius_icarus/board.c#L25

yath-eiSmart replied:
True, changed to devicetree property.

ycsin commented on drivers/sensor/wsen_hids/hids.c:
```diff
@@ -0,0 +1,296 @@
+
+/* Maps ODR configuration string (CONFIG_HIDS_ODR) to enum value used by HIDS driver. */
+static const struct str2odr hids_odr_string_list[] = {
+	{ "1", HIDS_odr1Hz },
```
please use designated initializer wherever possible, I think you dont need this if you used enum in the bindings

yath-eiSmart replied:
True, has been removed.

ycsin commented on drivers/sensor/wsen_hids/hids.c:
```diff
@@ -0,0 +1,296 @@
+static const struct sensor_value hids_odr_list[] = { { .val1 = 0, .val2 = 0 },
+						     { .val1 = 1, .val2 = 0 },
+						     { .val1 = 7, .val2 = 0 },
+						     { .val1 = 12, .val2 = 5 * 100000 } };
```
clang might be able to format this if you have comma

```suggestion
{ .val1 = 12, .val2 = 5 * 100000 }, };
```

otherwise it'd be nice if you can spend some time to format this manually

yath-eiSmart replied:
Done.

ycsin commented on drivers/sensor/wsen_hids/hids.c:
```diff
@@ -0,0 +1,296 @@
+		LOG_ERR("Failed to convert humidity sample.");
+		return -EIO;
+	}
+	if (WE_SUCCESS !=
```
it would be clearer to the reader if theres an extra line here to provide separation

yath-eiSmart replied:
Done.

ycsin commented on drivers/sensor/wsen_hids/hids.c:
```diff
@@ -0,0 +1,296 @@
+							  .trigger_set = hids_trigger_set,
+#endif
+							  .sample_fetch = hids_sample_fetch,
+							  .channel_get = hids_channel_get };
```
add comma and probably format manually

yath-eiSmart replied:
Done.

ycsin commented on drivers/sensor/wsen_hids/hids.c:
```diff
@@ -0,0 +1,296 @@
+		return -EINVAL;
+	}
+
+	if (WE_SUCCESS !=
```
hmm, why yoda notation here?

yath-eiSmart replied:
Overlooked this. Has been corrected.

galak commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -583,6 +583,13 @@ test_i2c_iis2iclx: iis2iclx@6a {
 
+test_i2c_hids: hids@5f {
+	compatible = "we,wsen-hids";
+	label = "WSEN-HIDS";
```
remove the label property, its not needed.

yath-eiSmart replied:
Done.

MaureenHelm reviewed (changes requested):
Thanks for the contribution!

Please split the changes into multiple commits (but keep them together in this one PR)
1. Add the hal module to west.yaml and related integration code that will be shared across multiple drivers
2. Add the new driver
3. Add the sample. Long term, I want to stop proliferating driver-specific samples in favor of more generalized driver class samples, see #47005. We haven't done this for humidity sensors yet, so I'm ok with this driver-specific sample for now.

ycsin commented on drivers/sensor/wsen_hids/hids_trigger.c:
```diff
@@ -0,0 +1,163 @@
+#ifdef CONFIG_HIDS_TRIGGER_OWN_THREAD
+static void hids_thread(struct hids_data *data)
+{
+	while (1) {
```
should probably use `while (true)`

yath-eiSmart replied:
Done.

ycsin commented on dts/bindings/sensor/we,wsen-hids-common.yaml:
```diff
@@ -0,0 +1,23 @@
+      required: false
+      description: |
+        Data-ready interrupt pin.
+
```
Not sure if this empty line is intentional?

mbolivar-nordic replied:
Agreed; this empty line should be removed.

yath-eiSmart replied:
Done.

ycsin commented on samples/sensor/wsen_hids/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+    harness: sensor
+    tags: sensors
+    depends_on: i2c hids
+    filter: dt_compat_enabled("we,wsen-hids")
```
empty line at the end of each file

yath-eiSmart replied:
Done.

ycsin commented on samples/sensor/wsen_hids/src/main.c:
```diff
@@ -0,0 +1,80 @@
+		return;
+	}
+
+	if (dev == NULL) {
```
I think `dev` probably wont be NULL, if it is null it wont compile, thanks to the `DEVICE_DT_GET_ONE`

ycsin replied:
you only need the `device_is_ready` to check if the driver is initialized successfully before using it

yath-eiSmart replied:
Done.

ycsin commented on samples/sensor/wsen_hids/src/main.c:
```diff
@@ -0,0 +1,80 @@
+	LOG_INF("HIDS device initialized.");
+
+	if (IS_ENABLED(CONFIG_HIDS_TRIGGER)) {
+		struct sensor_trigger trig = { .type = SENSOR_TRIG_DATA_READY,
+					       .chan = SENSOR_CHAN_ALL };
```
```suggestion
		struct sensor_trigger trig = {
			.type = SENSOR_TRIG_DATA_READY,
			.chan = SENSOR_CHAN_ALL,
		};
```

yath-eiSmart replied:
Done.

mbolivar-nordic commented on dts/bindings/vendor-prefixes.txt:
```diff
@@ -647,7 +647,7 @@ vxt	VXT Ltd
 waveshare	Waveshare Electronics
 wd	Western Digital Corp.
-we	Würth Elektronik GmbH.
+we	Würth Elektronik GmbH
```
This change should be in its own commit.

yath-eiSmart replied:
Ok, I've removed this change from the commit.

mbolivar-nordic commented on drivers/sensor/wsen_hids/Kconfig:
```diff
@@ -0,0 +1,51 @@
+# Copyright (c) 2022 Würth Elektronik eiSos GmbH & Co. KG
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig HIDS
```
IMO this should be WSEN_HIDS. "HIDS" is too easy to mistake for [human interface devices](https://en.wikipedia.org/wiki/Human_interface_device).

yath-eiSmart replied:
True. I had first named all the new WE drivers/identifiers similar to the existing ITDS driver, but using the full name including the "WSEN" prefix makes sense.

mbolivar-nordic commented on dts/bindings/sensor/we,wsen-hids-common.yaml:
```diff
@@ -0,0 +1,23 @@
+    odr:
+      type: string
+      required: false
+      default: "1"
```
You must explicitly document why you chose this default in the description.

Do **not** try to justify this via GitHub reply to this comment only: you **must** include the justification in the description in the binding.

See https://docs.zephyrproject.org/latest/build/dts/bindings.html#general-rules

yath-eiSmart replied:
Ok. I have actually removed the default value.

mbolivar-nordic reviewed (changes requested):
DT comments along with miscellany I happened to notice

galak commented on drivers/sensor/wsen_hids/Kconfig:
```diff
@@ -0,0 +1,51 @@
+menuconfig HIDS
+	bool "WSEN-HIDS humidity and temperature sensor"
+	depends on I2C || SPI
+	select HAS_WESENSORS
```
Add:
```
default y
depends on DT_HAS_WSEN_HIDS_ENABLED
```

Tweak `depends on I2C || SPI` to be:
```
+       depends on (I2C && $(dt_compat_on_bus,$(DT_COMPAT_WSEN_HIDS),i2c)) || \
+                  (SPI && $(dt_compat_on_bus,$(DT_COMPAT_WSEN_HIDS),spi))
```

galak replied:
@yath-eiSmart we are transitioning to the code I suggested.  Take a look at:

https://github.com/zephyrproject-rtos/zephyr/pull/48048

yath-eiSmart replied:
Ok, there was a "WE_" missing from `DT_COMPAT_WE_WSEN_HIDS` and `DT_HAS_WE_WSEN_HIDS_ENABLED`, that's why it didn't work when I tried it at first.
I've changed it accordingly.

MaureenHelm replied:
Still need to add selects for I2C and SPI

galak commented on tests/drivers/build_all/sensor/prj.conf:
```diff
@@ -35,6 +35,7 @@ CONFIG_ENS210=y
 CONFIG_FDC2X1X=y
 CONFIG_FXAS21002=y
 CONFIG_FXOS8700=y
+CONFIG_HIDS=y
```
With the Kconfig changes, you don't need to add `CONFIG_HIDS` here.

yath-eiSmart replied:
Done.

galak reviewed (dismissed):
Make changes to Kconfig

galak commented on samples/sensor/wsen_hids/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,15 @@
+	hids@5f {
+		compatible = "we,wsen-hids";
+		reg = <0x5f>;
+		label = "HIDS";
```
remove `label` property.

yath-eiSmart replied:
Done.

galak commented on samples/sensor/wsen_hids/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+  sample.sensor.wsen-hids:
+    harness: sensor
+    tags: sensors
+    depends_on: i2c hids
```
make this `depends on: i2c` only.  the `hids` part will get covered by the `filter`

yath-eiSmart replied:
Done.

yath-eiSmart committed:
modules: Added hal_wurthelektronik module.

Added hal_wurthelektronik module (contains drivers provided by
Würth Elektronik eiSos GmbH & Co. KG).

Signed-off-by: Yannick Thesen <Yannick.Thesen@we-online.de>

yath-eiSmart committed:
drivers: sensor: Added driver for the Würth Elektronik WSEN-HIDS sensor.

Added driver for the WSEN-HIDS humidity and temperature sensor.

Signed-off-by: Yannick Thesen <Yannick.Thesen@we-online.de>

yath-eiSmart committed:
drivers: sensor: Added sample for the Würth Elektronik WSEN-HIDS sensor.

Added sample for the WSEN-HIDS humidity and temperature sensor.

Signed-off-by: Yannick Thesen <Yannick.Thesen@we-online.de>

yath-eiSmart commented on samples/sensor/wsen_hids/src/main.c:
```diff
@@ -0,0 +1,75 @@
+
+void main(void)
+{
+	const struct device *dev = device_get_binding("HIDS");
```
Done.

yath-eiSmart commented on samples/sensor/wsen_hids/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,26 @@
+// 		label = "HIDS";
+// 		drdy-gpios = <&gpio0 24 GPIO_ACTIVE_HIGH>;
+// 	};
+// };
```
Thanks for the suggestions...done.

yath-eiSmart commented on modules/Kconfig.we:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2022 Würth Elektronik eiSos GmbH & Co. KG
+
+config HAS_WESENSORS
+	bool
```
Done.

yath-eiSmart commented on modules/Kconfig.we:
```diff
@@ -0,0 +1,6 @@
+# Würth Elektronik HAL config
+
+# Copyright (c) 2022 Würth Elektronik eiSos GmbH & Co. KG
```
Correct, changed to Apache-2.0.

yath-eiSmart commented on samples/sensor/wsen_hids/prj.conf:
```diff
@@ -0,0 +1,17 @@
+CONFIG_LOG=y
+# The following two parameters are required to enable float log support
+CONFIG_CBPRINTF_FP_SUPPORT=y
+CONFIG_LOG_MODE_DEFERRED=y
```
Done.

yath-eiSmart commented on drivers/sensor/wsen_hids/wsen_hids.c:
```diff
@@ -0,0 +1,277 @@
+	static const struct hids_config hids_config_##inst =                           \
+		COND_CODE_1(DT_INST_ON_BUS(inst, spi),                                 \
+			    (HIDS_CONFIG_SPI(inst)),                                   \
+			    (HIDS_CONFIG_I2C(inst)));                                  \
```
True. Changed it.

ycsin commented on drivers/sensor/wsen_hids/wsen_hids.c:
```diff
@@ -0,0 +1,277 @@
+	__ASSERT_NO_MSG(channel == SENSOR_CHAN_ALL);
+
+	if (HIDS_getRawValues(&data->sensor_interface, &rawHumidity, &rawTemp) != WE_SUCCESS) {
+		LOG_ERR("Failed to fetch data sample.");
```
```suggestion
		LOG_ERR("Failed to %s sample.", "fetch data");
```

yath-eiSmart replied:
Done.

ycsin commented on drivers/sensor/wsen_hids/wsen_hids.c:
```diff
@@ -0,0 +1,277 @@
+
+	if (HIDS_convertHumidity_uint16(&data->sensor_interface, rawHumidity, &data->humidity) !=
+	    WE_SUCCESS) {
+		LOG_ERR("Failed to convert humidity sample.");
```
```suggestion
		LOG_ERR("Failed to %s sample.", "convert humidity");
```

yath-eiSmart replied:
Done.

ycsin commented on drivers/sensor/wsen_hids/wsen_hids.c:
```diff
@@ -0,0 +1,277 @@
+
+	if (HIDS_convertTemperature_int16(&data->sensor_interface, rawTemp, &data->temperature) !=
+	    WE_SUCCESS) {
+		LOG_ERR("Failed to convert temperature sample.");
```
```suggestion
		LOG_ERR("Failed to %s sample.", "convert temperature");
```

This will save a few bytes in ROM

yath-eiSmart replied:
Done.

ycsin commented on drivers/sensor/wsen_hids/wsen_hids_trigger.c:
```diff
@@ -0,0 +1,163 @@
+static inline void hids_setup_drdy_interrupt(const struct device *dev, bool enable)
+{
+	const struct hids_config *cfg = dev->config;
+	unsigned int flags = enable ? GPIO_INT_EDGE_TO_ACTIVE : GPIO_INT_DISABLE;
```
type should be `gpio_flags_t`?

https://docs.zephyrproject.org/apidoc/latest/group__gpio__interface.html#ga5f5cb5e7dae6d58e072bb450af029d2e

ycsin commented on drivers/sensor/wsen_hids/wsen_hids_trigger.c:
```diff
@@ -0,0 +1,163 @@
+	const struct hids_config *cfg = dev->config;
+
+	if (trig->type != SENSOR_TRIG_DATA_READY) {
+		LOG_ERR("Unsupported sensor trigger");
```
same as above, would be more informative to print the `trig->type` arg

ycsin commented on drivers/sensor/wsen_hids/wsen_hids_trigger.c:
```diff
@@ -0,0 +1,163 @@
+
+	status = gpio_add_callback(cfg->gpio_drdy.port, &data->data_ready_cb);
+	if (status < 0) {
+		LOG_ERR("Could not set gpio callback.");
```
```suggestion
		LOG_ERR("Could not %s.", "set gpio callback");
```

ycsin commented on drivers/sensor/wsen_hids/wsen_hids_trigger.c:
```diff
@@ -0,0 +1,163 @@
+#if defined(CONFIG_WSEN_HIDS_TRIGGER_OWN_THREAD)
+	k_sem_init(&data->drdy_sem, 0, K_SEM_MAX_LIMIT);
+
+	k_thread_create(&data->thread, data->thread_stack, CONFIG_WSEN_HIDS_THREAD_STACK_SIZE,
```
```suggestion
	k_thread_create(&data->thread, data->thread_stack, K_KERNEL_STACK_SIZEOF(CONFIG_WSEN_HIDS_THREAD_STACK_SIZE),
```

yath-eiSmart replied:
Thanks for the suggestion. I don't really get the usage of `K_KERNEL_STACK_SIZEOF` here though. Should it be used with `data->thread_stack` as argument instead?

On the other hand - all the other sensor drivers seem to use a `CONFIG_*_THREAD_STACK_SIZE` variable directly as input argument to `k_thread_create`.

ycsin replied:
I remember that the macro will help to make sure that the stack is aligned, would love your input @andyross

ycsin commented on samples/sensor/wsen_hids/prj.conf:
```diff
@@ -0,0 +1,16 @@
+###### Basic configuration required for using sensors ######
+CONFIG_I2C=y
+# CONFIG_SPI=y
```
probably nice to comment why SPI is disabled, probably can encourage user to have a look at relevate file

yath-eiSmart replied:
I simply removed the line.

ycsin commented on samples/sensor/wsen_hids/prj.conf:
```diff
@@ -0,0 +1,16 @@
+
+###### Configuration of WSEN-HIDS ######
+CONFIG_WSEN_HIDS_TRIGGER_NONE=y
+# CONFIG_WSEN_HIDS_TRIGGER_GLOBAL_THREAD=y
```
I know your intention but I'd rather keep the sample simple (not a blocker, no strong feeling)

yath-eiSmart replied:
Ok, I removed all lines that are not required.

ycsin commented on samples/sensor/wsen_hids/src/main.c:
```diff
@@ -0,0 +1,77 @@
+
+	while (!IS_ENABLED(CONFIG_WSEN_HIDS_TRIGGER)) {
+		process_sample(dev);
+		k_sleep(K_MSEC(2000));
```
can probably just do `k_msleep` for msec

yath-eiSmart replied:
Done.

ycsin commented on west.yml:
```diff
@@ -139,6 +139,13 @@ manifest:
         - hal
+    - name: hal_we
+      revision: bb79ef1cc46d00058e89b48ad36cca8efebbbf0e
+      # Explicit URL (temporary)
```
temporary?

yath-eiSmart replied:
Yes - the "hal_we" repository is currently hosted outside the Zephyr project, until the new module repository has been created.

See https://github.com/zephyrproject-rtos/zephyr/issues/47021#issuecomment-1190499840

MaureenHelm replied:
The new module repo has been created

yath-eiSmart replied:
I don't have access to the repo yet - so I made all the requested changes except for this one.

ycsin reviewed (changes requested):
had another look

yath-eiSmart commented on drivers/sensor/wsen_hids/wsen_hids.c:
```diff
@@ -0,0 +1,277 @@
+			 enum sensor_attribute attr, const struct sensor_value *val)
+{
+	if (chan != SENSOR_CHAN_ALL) {
+		LOG_WRN("attr_set() not supported on this channel.");
```
Done.

Btw: This was also a snippet taken from existing sensor drivers. So most drivers don't print the channel, some do.

yath-eiSmart commented on drivers/sensor/wsen_hids/wsen_hids.c:
```diff
@@ -0,0 +1,277 @@
+		return -ENOTSUP;
+	}
+
+	return 0;
```
Done.

mah-eiSmart commented:
Hello @ycsin 
any progress in review? What's the current state?
Thx in advance,
Matthias

MaureenHelm commented on drivers/sensor/wsen_hids/wsen_hids.c:
```diff
@@ -0,0 +1,272 @@
+	.channel_get = hids_channel_get,
+};
+
+int hids_init(const struct device *dev)
```
```suggestion
static int hids_init(const struct device *dev)
```

MaureenHelm replied:
This still needs to be addressed

yath-eiSmart replied:
Done.

MaureenHelm commented on drivers/sensor/wsen_hids/wsen_hids.c:
```diff
@@ -0,0 +1,272 @@
+	HIDS_getDefaultInterface(&data->sensor_interface);
+	data->sensor_interface.interfaceType = interface_type;
+	if (data->sensor_interface.interfaceType == WE_i2c) {
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+		data->sensor_interface.handle = (void *)&config->bus_cfg.i2c;
+#endif
+	} else {
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+		data->sensor_interface.handle = (void *)&config->bus_cfg.spi;
+#endif
```
This conditional logic isn't quite right - you want to test if *this* instance is on an I2C or SPI bus, not *any* instance.

MaureenHelm replied:
This still needs to be addressed

yath-eiSmart replied:
@MaureenHelm Thanks for the feedback. However, I think in this case the logic is in fact correct, as there is a surrounding if-statement that checks if the instance is on an I2C or SPI bus. `DT_ANY_INST_ON_BUS_STATUS_OKAY` is used, as the corresponding field of the `config->bus_cfg` union is included in the compilation based on the same logic.

yath-eiSmart replied:
Yes, that does make sense. Changed it.

mah-eiSmart commented:
Hello @MaureenHelm ,

thanks for the feedback. I edited the code and updated the pull request to the most recent version.

BR Matthias

MaureenHelm commented on drivers/sensor/wsen_hids/wsen_hids.c:
```diff
@@ -0,0 +1,272 @@
+static int hids_sample_fetch(const struct device *dev, enum sensor_channel channel)
+{
+	struct hids_data *data = dev->data;
+	int16_t rawHumidity;
+	int16_t rawTemp;
```
Please don't use camel case

MaureenHelm reviewed (changes requested):
Please rebase and remove the merge commit

MaureenHelm reviewed (commented):
One nit, otherwise LGTM

MaureenHelm reviewed (approved):
Looks great, thank you.

Is there a reason this is still labeled `DNM`?

mah-eiSmart commented:
> Looks great, thank you.
> 
> Is there a reason this is still labeled `DNM`?

I don't think so. Merging can be done in my opinion.

BTW: I'm responsible for this contribution, now, as @yath-eiSmart is out of office.

BR Matthias

fabiobaltieri commented:
Merged, mind you there were a couple of minor open comments still open on opaque types and logging messages without information about the fail condition, I think they were still relevant, feel free to send a followup @yath-eiSmart.
